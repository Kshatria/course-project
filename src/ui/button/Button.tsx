import type { ButtonProps } from './Button.types';
import { clsx } from 'clsx';
import styles from './Button.module.css';

const Button = ({ color, disabled, text, type = 'button', onClick }: ButtonProps) => {
  const btnClasses = clsx(styles.button, {
    [styles.disabled]: disabled,
    [styles.primary]: color === 'Primary',
    [styles.secondary]: color === 'Secondary',
  });

  return (
    <button
      className={btnClasses}
      type={type}
      onClick={disabled ? undefined : onClick}
      disabled={disabled}
    >
      {text}
    </button>
  );
};

export { Button, type ButtonProps };
