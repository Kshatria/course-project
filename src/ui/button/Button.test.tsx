import { fireEvent, render, screen } from '@testing-library/react';
import { Button, type ButtonProps } from './Button';

describe('Button Component', () => {
  const defaultProps: ButtonProps = {
    text: 'Click me',
    onClick: jest.fn(),
  };

  it('matches snapshot (testing-library)', () => {
    const { asFragment } = render(<Button {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches disabled snapshot', () => {
    const { asFragment } = render(<Button {...defaultProps} disabled />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches primary snapshot', () => {
    const { asFragment } = render(<Button {...defaultProps} color="Primary" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders button with correct text', () => {
    render(<Button {...defaultProps} />);
    expect(screen.getByText(defaultProps.text)).toBeInTheDocument();
  });

  it('calls onClick handler when clicked', () => {
    render(<Button {...defaultProps} />);
    fireEvent.click(screen.getByText(defaultProps.text));
    expect(defaultProps.onClick).toHaveBeenCalledTimes(1);
  });

  it('has default type "button"', () => {
    render(<Button {...defaultProps} />);
    expect(screen.getByText(defaultProps.text)).toHaveAttribute('type', 'button');
  });

  it('applies correct type attribute when provided', () => {
    const type = 'submit';
    render(<Button {...defaultProps} type={type} />);
    expect(screen.getByText(defaultProps.text)).toHaveAttribute('type', type);
  });

  it('is disabled when disabled prop is true', () => {
    render(<Button {...defaultProps} disabled={true} />);
    const button = screen.getByText(defaultProps.text);
    expect(button).toBeDisabled();
    expect(button).toHaveAttribute('disabled');
    expect(button).toHaveClass('disabled');
  });

  it('does not call onClick when disabled', () => {
    const onClick = jest.fn();
    render(<Button {...defaultProps} disabled={true} onClick={onClick} />);
    const button = screen.getByText(defaultProps.text);

    fireEvent.click(button);
    expect(onClick).not.toHaveBeenCalled();
  });

  describe('CSS Classes', () => {
    it('has default button class', () => {
      render(<Button {...defaultProps} />);
      const button = screen.getByText(defaultProps.text);
      expect(button).toHaveClass('button');
    });

    it('has disabled class when disabled', () => {
      render(<Button {...defaultProps} disabled={true} />);
      const button = screen.getByText(defaultProps.text);
      expect(button).toHaveClass('disabled');
    });

    it('has primary class when color is Primary', () => {
      render(<Button {...defaultProps} color="Primary" />);
      const button = screen.getByText(defaultProps.text);
      expect(button).toHaveClass('primary');
    });

    it('has secondary class when color is Secondary', () => {
      render(<Button {...defaultProps} color="Secondary" />);
      const button = screen.getByText(defaultProps.text);
      expect(button).toHaveClass('secondary');
    });
  });
});
