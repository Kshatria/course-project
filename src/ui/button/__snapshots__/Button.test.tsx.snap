// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component matches disabled snapshot 1`] = `
<DocumentFragment>
  <button
    class="button disabled"
    disabled=""
    type="button"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`Button Component matches primary snapshot 1`] = `
<DocumentFragment>
  <button
    class="button primary"
    type="button"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`Button Component matches snapshot (testing-library) 1`] = `
<DocumentFragment>
  <button
    class="button"
    type="button"
  >
    Click me
  </button>
</DocumentFragment>
`;
